File: SixUiViews/app/src/main/java/cn/six/open/view/rv/OneAdapter/OneAdapter.java
package cn.six.open.view.rv.OneAdapter;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by songzhw on 2016-08-13
 */
public abstract class OneAdapter<T> extends RecyclerView.Adapter<RvViewHolder> {
    public int layoutResId;
    public List<T> data;

    public OneAdapter(int layoutResId) {
        this.layoutResId = layoutResId;
        data = new ArrayList<>();
    }

    public OneAdapter(int layoutResId, List<T> data){
        this.layoutResId = layoutResId;
        this.data = data;
    }

    @Override
    public int getItemCount() {
        return data == null ? 0 : data.size();
    }

    @Override
    public RvViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RvViewHolder vh = RvViewHolder.createViewHolder(parent, layoutResId);
        return vh;
    }

    @Override
    public void onBindViewHolder(RvViewHolder holder, int position) {
        if(data != null && data.size() > position){
            apply(holder, data.get(position), position);
        }
    }


    protected abstract void apply(RvViewHolder vh, T t, int position);

}

File: SixUiViews/lib/src/main/java/ca/six/views/rv/single/OneAdapter.kt
package ca.six.views.rv.single


import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import ca.six.views.rv.RvViewHolder
import java.util.*

abstract class OneAdapter<T> : RecyclerView.Adapter<RvViewHolder> {
    var layoutResId: Int = 0
    var data: List<T>? = null

    constructor(layoutResId: Int) {
        this.layoutResId = layoutResId
        data = ArrayList()
    }

    constructor(layoutResId: Int, data: List<T>) {
        this.layoutResId = layoutResId
        this.data = data
    }

    override fun getItemCount(): Int {
        return if (data == null) 0 else data!!.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RvViewHolder {
        return RvViewHolder.createViewHolder(parent, layoutResId)
    }

    override fun onBindViewHolder(holder: RvViewHolder, position: Int) {
        if (data != null && data!!.size > position) {
            apply(holder, data!![position], position)
        }
    }


    protected abstract fun apply(vh: RvViewHolder, value: T, position: Int)

}